code/rd-vanilla/G2_misc.cpp

```diff
@@ -1760,162 +1770,212 @@ void *G2_FindSurface(const model_s *mod, int index, int lod)
 	return (void *)current;
 }
 
+#if 1
 #define SURFACE_SAVE_BLOCK_SIZE	sizeof(surfaceInfo_t)
 #define BOLT_SAVE_BLOCK_SIZE sizeof(boltInfo_t)
 #define BONE_SAVE_BLOCK_SIZE sizeof(boneInfo_t)
+#else
+const auto SURFACE_SAVE_BLOCK_SIZE = static_cast<int>(sizeof(SgSurfaceInfo));
+const auto BOLT_SAVE_BLOCK_SIZE = static_cast<int>(sizeof(SgBoltInfo));
+const auto BONE_SAVE_BLOCK_SIZE = static_cast<int>(sizeof(SgBoneInfo));
+#endif
 
-void G2_SaveGhoul2Models(CGhoul2Info_v &ghoul2)
+
+void G2_SaveGhoul2Models(
+	CGhoul2Info_v& ghoul2)
 {
-	char *pGhoul2Data = NULL;
-	int   iGhoul2Size = 0;
+	ojk::SavedGameHelper saved_game(
+		::ri.saved_game);
+
+	saved_game.reset_buffer();
 
 	// is there anything to save?
-	if (!ghoul2.IsValid()||!ghoul2.size())
+	if (!ghoul2.IsValid() || ghoul2.size() == 0)
 	{
-		ri.SG_Append(INT_ID('G','H','L','2'),&pGhoul2Data, 4);	//write out a zero buffer
-		return;
-	}
+		const int zero_size = 0;
 
-	// this one isn't a define since I couldn't work out how to figure it out at compile time
-	const int ghoul2BlockSize = (intptr_t)&ghoul2[0].BSAVE_END_FIELD - (intptr_t)&ghoul2[0].BSAVE_START_FIELD;
+#ifdef JK2_MODE
+		saved_game.write<int32_t>(
+			zero_size);
 
-	// add in count for number of ghoul2 models
-	iGhoul2Size += 4;
-	// start out working out the total size of the buffer we need to allocate
-	for (int i=0; i<ghoul2.size();i++)
-	{
-		iGhoul2Size += ghoul2BlockSize;
-		// add in count for number of surfaces
-		iGhoul2Size += 4;
-		iGhoul2Size += (ghoul2[i].mSlist.size() * SURFACE_SAVE_BLOCK_SIZE);
-		// add in count for number of bones
-		iGhoul2Size += 4;
-		iGhoul2Size += (ghoul2[i].mBlist.size() * BONE_SAVE_BLOCK_SIZE);
-		// add in count for number of bolts
-		iGhoul2Size += 4;
-		iGhoul2Size += (ghoul2[i].mBltlist.size() * BOLT_SAVE_BLOCK_SIZE);
-	}
+		saved_game.write_chunk_and_size<int32_t>(
+			INT_ID('G', 'L', '2', 'S'),
+			INT_ID('G', 'H', 'L', '2'));
+#else
+		saved_game.write_chunk<int32_t>(
+			INT_ID('G', 'H', 'L', '2'),
+			zero_size); //write out a zero buffer
+#endif // JK2_MODE
 
-	// ok, we should know how much space we need now
-	pGhoul2Data = (char*)R_Malloc(iGhoul2Size, TAG_GHOUL2, qfalse);
+		return;
+	}
 
-	// now lets start putting the data we care about into the buffer
-	char *tempBuffer = pGhoul2Data;
 
 	// save out how many ghoul2 models we have
-	*(int *)tempBuffer = ghoul2.size();
-	tempBuffer +=4;
+	auto model_count = ghoul2.size();
```

Comment:
Remove the auto here please! Cast to int or size_t and then remove all the decltype expressions below. I think it makes things easier to read.